plat/kvm/x86/acpi.c

```diff
@@ -0,0 +1,202 @@
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ */
+
+#include <uk/print.h>
+#include <string.h>
+#include <x86/acpi/acpi.h>
+#include <x86/acpi/madt.h>
+#include <uk/assert.h>
+
+__u8 acpi_version;
+struct RSDPDescriptor *rsdp;
+struct RSDPDescriptor20 *rsdp2;
+struct RSDT *rsdt;
+struct XSDT *xsdt;
+struct MADT *madt;
+
+static inline __u8 detect_acpi_version(void *ptr)
+{
+	struct RSDPDescriptor *rsdp = (struct RSDPDescriptor *)ptr;
+
+	UK_ASSERT(ptr);
+
+	return rsdp->Revision;
+}
+
+static inline __s8 rsdp_checksum(__u64 *ptr)
+{
+	__u32 checksum = 0;
+	__u8 *i;
+
+	for (i = (__u8 *)ptr; i < (__u8 *)ptr + sizeof(struct RSDPDescriptor);
+	     i++)
+		checksum += *i;
+
+	return (__u8)checksum == 0 ? 0 : -1;
+}
+
+/*
+ * Not yet implemented, because I have not yet detected ACPI version 2 or
+ * greater.
+ */
+
+static inline __s8 rsdp2_checksum(__u64 *ptr)
+{
+	return 0;
+}
+
+/*
+ * Do checksum for any ACPI table, except RSDP.
+ */
+
+static inline __s8 acpi_checksum(struct ACPISDTHeader *h)
+{
+	__u8 checksum = 0;
+	__u32 i;
+
+	for (i = 0; i < h->Length; i++)
+		checksum += ((__u8 *)h)[i];
+
+	return checksum == 0 ? 0 : -1;
+}
+
+/*
+ * Find the ACPI Root System Descriptor Table (RSDT), and check if it's valid.
+ * For this, the Root System Descriptor Pointer (RSDP) must be found, in the
+ * physical memory area 0xe0000 -> 0xfffff.
+ */
+
+__s8 find_rsdt(void)
+{
+	__uptr start_addr = 0xe0000, end_addr = 0xfffff;
+	__uptr i;
+	__u64 *ptr;
+
+	for (i = start_addr; i < end_addr; i += 16) {
+		ptr = (__u64 *)i;
+		if (!memcmp(ptr, "RSD PTR ", 8)) {
+			uk_pr_info("Found RSDP at %p\n", ptr);
+			break;
+		}
+	}
+
+	uk_pr_info("ACPI version detected: ");
+	if (detect_acpi_version(ptr) == 0) {
+		uk_pr_info("1.0\n");
+		if (rsdp_checksum(ptr) != 0) {
+			uk_pr_err("Invalid RSDP\n");
+			return -1;
+		}
+
+		acpi_version = 1;
+		rsdp = (struct RSDPDescriptor *)ptr;
+		rsdt = (struct RSDT *)rsdp->RsdtAddress;
+
+		if (acpi_checksum(&rsdt->h) != 0) {
+			uk_pr_err("Invalid RSDT\n");
+			rsdt = NULL;
+			return -1;
+		}
+
+		uk_pr_info("RSDT present at %p\n", rsdt);
+	} else {
+		/*
+		 * TODO: add support for ACPI version 2 or greater, when/ if it
+		 * will be detected
+		 */
+
+		uk_pr_info(">= 2\n");
+		uk_pr_err("ACPI version not yet supported!\n");
+
+		return -1;
+	}
+#ifdef UK_DEBUG
+	list_acpi_tables();
+#endif /* UK_DEBUG */
+
+	return 0;
+}
+
+/*
+ * Print the detected ACPI tables, for debug.
+ */
+
+void list_acpi_tables(void)
+{
```

Comment:
Agree. Would guard it and make it static (i.e., not export it).

Concerning coherence with the coding style, I also suggest to remove the `rsdt` test as commented by Simon for `find_madt` and put the variable definitions at the top. Then add an `UK_ASSERT` and after that compute the number of entries. The same applies to `find_madt`.

```C
#ifdef UK_DEBUG
void list_acpi_tables(void)
{
	int i, entries;
	struct ACPISDTHeader *h;

	UK_ASSERT(rsdt);
 	entries = (rsdt->h.Length - sizeof(rsdt->h)) / 4;

	uk_pr_debug("%d ACPI tables found\n", entries);

	for (i = 0; i < entries; i++) {
		h = (struct ACPISDTHeader *)rsdt->PointerToOtherSDT[i];
		uk_pr_debug("%p: %.4s\n", h, h->Signature);
	}
}
#endif
```

Also consider to use a macro for the computation of the number of entries, since you use it also in `find_madt`.